File: Programmers/LEVEL_2/c++/프린터.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    queue<pair<int, int> > q;
    priority_queue<int> pq;

    for (int i = 0; i < priorities.size(); i++)
    {
        q.push(make_pair(i, priorities[i]));
        pq.push(priorities[i]);
    }

    int res = 0;
    while (!pq.empty())
    {
        int i = q.front().first;
        int value = q.front().second;
        q.pop();

        if (value == pq.top())
        {
            pq.pop();
            answer += 1;
            if (i == location) break;
        }
        else
        {
            q.push(make_pair(i, value));
        }
    }

    return answer;
}

File: Programmers/LEVEL_2/프린터.js
function solution(priorities, location) {
  var answer = 0;

  // map(요소, 인덱스, 배열)
  let arr = priorities.map((priority, index) => {
    return { priority: priority, index: index };
  });

  let queue = [];
  while (arr.length > 0) {
    let temp = arr.shift();
    let check = arr.some((e) => e.priority > temp.priority); // 각 요소가 모두 통과하면 true
    if (check) {
      arr.push(temp);
    } else {
      // false면 가장 크다는 뜻이니까 queue에 넣자.
      queue.push(temp);
    }
  }
  answer = queue.findIndex((q) => q.index == location) + 1;

  return answer;
}
